sync/awsdatabricks.py

```diff
@@ -1017,71 +1047,205 @@ def _s3_contents_have_all_rollover_logs(contents: List[dict], run_end_time_secon
+    #  we make sure to re-strip our final Prefix
+    # TODO - using the spark_context_id might be good here, as we do in the DBFS logic
+    prefix = f"{base_filepath}/eventlog/{cluster_id}".strip("/")
+
+    logger.info(f"Looking for eventlogs at location: {prefix}")
+
+    contents = s3.list_objects_v2(Bucket=bucket, Prefix=prefix).get("Contents")
+    run_end_time_seconds = run_end_time_millis / 1000
+    poll_num_attempts = 0
+    poll_max_attempts = 20  # 5 minutes / 15 seconds = 20 attempts
+    while (
+        not _s3_contents_have_all_rollover_logs(contents, run_end_time_seconds)
+        and poll_num_attempts < poll_max_attempts
+    ):
+        if poll_num_attempts > 0:
+            logger.info(
+                f"No or incomplete event log data detected - attempting again in {poll_duration_seconds} seconds"
+            )
+            sleep(poll_duration_seconds)
 
-        cluster_id = cluster_description["cluster_id"]
-        s3 = boto.client("s3")
+        contents = s3.list_objects_v2(Bucket=bucket, Prefix=prefix).get("Contents")
+        poll_num_attempts += 1
+
+    if contents:
+        eventlog_zip = io.BytesIO()
+        eventlog_zip_file = zipfile.ZipFile(eventlog_zip, "a", zipfile.ZIP_DEFLATED)
+
+        for content in contents:
+            obj = s3.get_object(Bucket=bucket, Key=content["Key"])
+            eventlog_zip_file.writestr(content["Key"].split("/")[-1], obj["Body"].read())
+
+        eventlog_zip_file.close()
+
+        return Response(result=eventlog_zip.getvalue())
+
+    return Response(
+        error=DatabricksError(
+            message=f"No eventlog found at location - {bucket}/{base_filepath} - after {poll_num_attempts * poll_duration_seconds} seconds"
+        )
+    )
+
+
+def _get_eventlog_from_dbfs(
+    cluster_id: str,
+    spark_context_id: str,
+    base_filepath: str,
+    run_end_time_millis: int,
+    poll_duration_seconds: int,
+):
+    prefix = f"dbfs:/{base_filepath}/eventlog/"
+
+    root_dir = get_default_client().list_dbfs_directory(prefix)
+
+    eventlog_files = [f for f in root_dir["files"] if f["is_dir"]]
+    matching_subdirectory = None
+
+    # For shared clusters, we may find multiple subdirectories under the same root path, in the format -
+    #     {cluster_id}_{driver_ip_address}
+    # Since DBFS gives us no good filtering mechanism for this, and there isn't always an easy way to get
+    #  the driver's IP address for a Run, we can list the subdirectories and look for one containing a path
+    #  that ends with this cluster's `spark_context_id`
+    while eventlog_files and not matching_subdirectory:
+        eventlog_file_metadata = eventlog_files.pop()
+        path = eventlog_file_metadata["path"]
+
+        subdir = get_default_client().list_dbfs_directory(path)
+
+        subdir_files = subdir["files"]
+        matching_subdirectory = next(
+            (subfile for subfile in subdir_files if spark_context_id in subfile["path"]),
+            None,
+        )
```

Comment:
Use the `spark_context_id` in order to figure out the correct sub-directory to be looking in

Reply:
Interesting that it's not used in S3 eventlog paths

Reply:
It is, if you look at the full path in S3, but we just don't use it (yet)